cmd/katib-controller/v1beta1/main.go

```diff
@@ -127,36 +129,61 @@ func main() {
 		os.Exit(1)
 	}
 
-	log.Info("Registering Components.")
+	// Create a webhook server.
+	hookServer := webhook.NewServer(webhook.Options{
+		Port:    *initConfig.ControllerConfig.WebhookPort,
+		CertDir: consts.CertDir,
+	})
 
-	// Setup all Controllers
-	log.Info("Setting up controller.")
-	if err := controller.AddToManager(mgr); err != nil {
-		log.Error(err, "Unable to register controllers to the manager")
-		os.Exit(1)
-	}
+	ctx := signals.SetupSignalHandler()
+	certsReady := make(chan struct{})
 
-	log.Info("Setting up webhooks.")
-	if err := webhook.AddToManager(mgr, *initConfig.ControllerConfig.WebhookPort); err != nil {
-		log.Error(err, "Unable to register webhooks to the manager")
-		os.Exit(1)
+	if initConfig.CertGeneratorConfig.Enable {
+		if err = cert.AddToManager(mgr, initConfig.CertGeneratorConfig, certsReady); err != nil {
```

Comment:
Adding a cert-generator to the manager has a couple of advantages:

1. It makes it easy to handle errors in the manager when the cert-generator fails.
2. if we don't add a cert-generator to the manager, we must implement processes to wait for starting the manager to wait for starting client cache server on the cert-generator side. The client cache server is started by `mgr.Start()`.

Also, as future work, we can improve the cert-generator so that rotating the certs at regular intervals.

Reply:
Please can you explain how do we use client cache in the cert-generator ?
Do you mean it improves process of getting K8s resources on the cert generator side ?
Since we store resources in the cache (e.g. admission webhooks), we don't need to invoke Kubernetes API Server to get it.

Reply:
> Please can you explain how do we use client cache in the cert-generator ?

The cert-generator gets the K8s client from manager like this:

https://github.com/kubeflow/katib/blob/ed91adbe975b509ba67cf24094c5f0265af1670f/pkg/cert-generator/v1beta1/generator.go#L90

The client uses a cache for READ operations, and uses direct client (call kube-apiserver) for WRITE operations.

https://github.com/kubernetes-sigs/controller-runtime/blob/7f0c6dc440f334849b25866dd9990ac1a5a777c2/pkg/cluster/cluster.go#L53-L57

So, we need to start the cache server.

Reply:
Makes sense, thanks!